File: src/clients/javascript/SkydropxClient.js
const axios = require('axios');

/**
 * Cliente profesional para la API de Skydropx
 * 
 * @class SkydropxClient
 * @description Cliente completo para interactuar con todos los endpoints de la API de Skydropx
 * @author Skydropx Community
 * @version 1.0.0
 * 
 * @example
 * const client = new SkydropxClient({
 *   clientId: 'your_client_id',
 *   clientSecret: 'your_client_secret',
 *   environment: 'sandbox'
 * });
 * 
 * await client.authenticate();
 * const quotation = await client.createQuotation({...});
 */
class SkydropxClient {
  /**
   * @param {Object} config - Configuración del cliente
   * @param {string} config.clientId - Client ID de Skydropx
   * @param {string} config.clientSecret - Client Secret de Skydropx
   * @param {string} [config.environment='sandbox'] - Ambiente: 'sandbox' o 'production'
   * @param {number} [config.timeout=30000] - Timeout de las peticiones en ms
   * @param {number} [config.retryAttempts=3] - Número de reintentos en caso de error
   */
  constructor(config) {
    this.clientId = config.clientId;
    this.clientSecret = config.clientSecret;
    this.environment = config.environment || 'sandbox';
    this.timeout = config.timeout || 30000;
    this.retryAttempts = config.retryAttempts || 3;
    
    // URL base según ambiente
    this.baseURL = 'https://app.skydropx.com';
    
    this.accessToken = null;
    this.tokenExpiresAt = null;
    
    // Configurar axios
    this.httpClient = axios.create({
      baseURL: this.baseURL,
      timeout: this.timeout,
      headers: {
        'Content-Type': 'application/json'
      }
    });
    
    // Interceptor para manejar renovación automática de token
    this.httpClient.interceptors.request.use(
      async (config) => {
        if (this.shouldRenewToken()) {
          await this.authenticate();
        }
        
        if (this.accessToken && !config.url.includes('/oauth/token')) {
          config.headers.Authorization = `Bearer ${this.accessToken}`;
        }
        
        return config;
      },
      (error) => Promise.reject(error)
    );
    
    // Interceptor para manejar errores
    this.httpClient.interceptors.response.use(
      (response) => response,
      async (error) => {
        if (error.response?.status === 401 && !error.config._retry) {
          error.config._retry = true;
          await this.authenticate();
          error.config.headers.Authorization = `Bearer ${this.accessToken}`;
          return this.httpClient(error.config);
        }
        return Promise.reject(error);
      }
    );
  }
  
  /**
   * Verifica si el token debe renovarse
   * @private
   */
  shouldRenewToken() {
    if (!this.accessToken || !this.tokenExpiresAt) return true;
    
    // Renovar 5 minutos antes de expirar
    const now = Date.now();
    const fiveMinutes = 5 * 60 * 1000;
    return now >= (this.tokenExpiresAt - fiveMinutes);
  }
  
  // ==================== AUTENTICACIÓN ====================
  
  /**
   * Obtiene un token de acceso
   * @returns {Promise<Object>} Información del token
   * @throws {Error} Si las credenciales son inválidas
   * 
   * @example
   * await client.authenticate();
   */
  async authenticate() {
    try {
      const response = await this.httpClient.post('/api/v1/oauth/token', {
        client_id: this.clientId,
        client_secret: this.clientSecret,
        grant_type: 'client_credentials'
      });
      
      this.accessToken = response.data.access_token;
      
      // Calcular tiempo de expiración (expires_in en segundos)
      const expiresIn = response.data.expires_in * 1000; // Convertir a ms
      this.tokenExpiresAt = Date.now() + expiresIn;
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al autenticar', error);
    }
  }
  
  /**
   * Revoca el token actual
   * @returns {Promise<void>}
   */
  async revokeToken() {
    try {
      await this.httpClient.post('/api/v1/oauth/revoke', {
        client_id: this.clientId,
        client_secret: this.clientSecret,
        token: this.accessToken,
        token_type_hint: 'access_token'
      });
      
      this.accessToken = null;
      this.tokenExpiresAt = null;
    } catch (error) {
      throw this.handleError('Error al revocar token', error);
    }
  }
  
  /**
   * Obtiene información del token actual
   * @returns {Promise<Object>} Detalles del token
   */
  async introspectToken() {
    try {
      const response = await this.httpClient.post('/api/v1/oauth/introspect', {
        client_id: this.clientId,
        client_secret: this.clientSecret,
        token: this.accessToken,
        token_type_hint: 'access_token'
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al inspeccionar token', error);
    }
  }
  
  // ==================== COTIZACIONES ====================
  
  /**
   * Crea una cotización
   * @param {Object} quotationData - Datos de la cotización
   * @returns {Promise<Object>} Cotización creada
   * 
   * @example
   * const quotation = await client.createQuotation({
   *   address_from: {
   *     country_code: 'MX',
   *     postal_code: '64000',
   *     area_level1: 'Nuevo León',
   *     area_level2: 'Monterrey',
   *     area_level3: 'Centro'
   *   },
   *   address_to: {
   *     country_code: 'MX',
   *     postal_code: '01000',
   *     area_level1: 'Ciudad de México',
   *     area_level2: 'Álvaro Obregón',
   *     area_level3: 'Santa Fe'
   *   },
   *   packages: [{
   *     weight: 2.5,
   *     length: 30,
   *     width: 20,
   *     height: 15
   *   }]
   * });
   */
  async createQuotation(quotationData) {
    try {
      const response = await this.httpClient.post('/api/v1/quotations', {
        quotation: quotationData
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al crear cotización', error);
    }
  }
  
  /**
   * Obtiene una cotización por ID
   * @param {string} quotationId - ID de la cotización
   * @returns {Promise<Object>} Detalles de la cotización
   */
  async getQuotation(quotationId) {
    try {
      const response = await this.httpClient.get(`/api/v1/quotations/${quotationId}`);
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener cotización', error);
    }
  }
  
  // ==================== ENVÍOS ====================
  
  /**
   * Crea un envío
   * @param {Object} shipmentData - Datos del envío
   * @returns {Promise<Object>} Envío creado
   * 
   * @example
   * const shipment = await client.createShipment({
   *   rate_id: 'rate_id_from_quotation',
   *   address_from: {
   *     name: 'Juan Pérez',
   *     company: 'Mi Empresa',
   *     street1: 'Av. Principal 123',
   *     phone: '8112345678',
   *     email: 'juan@example.com',
   *     reference: 'Edificio azul'
   *   },
   *   address_to: {
   *     name: 'María García',
   *     street1: 'Calle 456',
   *     phone: '5587654321',
   *     email: 'maria@example.com'
   *   }
   * });
   */
  async createShipment(shipmentData) {
    try {
      const response = await this.httpClient.post('/api/v1/shipments', {
        shipment: shipmentData
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al crear envío', error);
    }
  }
  
  /**
   * Obtiene todos los envíos (paginados)
   * @param {number} [page=1] - Número de página
   * @returns {Promise<Object>} Lista de envíos
   */
  async getShipments(page = 1) {
    try {
      const response = await this.httpClient.get('/api/v1/shipments', {
        params: { page }
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener envíos', error);
    }
  }
  
  /**
   * Obtiene un envío por ID
   * @param {string} shipmentId - ID del envío
   * @returns {Promise<Object>} Detalles del envío
   */
  async getShipment(shipmentId) {
    try {
      const response = await this.httpClient.get(`/api/v1/shipments/${shipmentId}`);
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener envío', error);
    }
  }
  
  /**
   * Cancela un envío
   * @param {string} shipmentId - ID del envío
   * @param {string} reason - Razón de cancelación
   * @returns {Promise<Object>} Confirmación de cancelación
   */
  async cancelShipment(shipmentId, reason = 'Cancelado por el usuario') {
    try {
      const response = await this.httpClient.post(
        `/api/v1/shipments/${shipmentId}/cancellations`,
        { reason, shipment_id: shipmentId }
      );
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al cancelar envío', error);
    }
  }
  
  /**
   * Protege un envío
   * @param {string} shipmentId - ID del envío
   * @param {number} declaredValue - Valor declarado del envío
   * @returns {Promise<Object>} Detalles de la protección
   */
  async protectShipment(shipmentId, declaredValue) {
    try {
      const response = await this.httpClient.post(
        `/api/v1/shipments/${shipmentId}/protect`,
        {
          protect: {
            declared_value: declaredValue,
            shipment_id: shipmentId
          }
        }
      );
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al proteger envío', error);
    }
  }
  
  /**
   * Rastrea un envío
   * @param {string} trackingNumber - Número de rastreo
   * @param {string} carrierName - Nombre de la paquetería (fedex, dhl, etc)
   * @returns {Promise<Object>} Eventos de rastreo
   * 
   * @example
   * const tracking = await client.trackShipment('123456789', 'fedex');
   */
  async trackShipment(trackingNumber, carrierName) {
    try {
      const response = await this.httpClient.get('/api/v1/shipments/tracking', {
        params: {
          tracking_number: trackingNumber,
          carrier_name: carrierName
        }
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al rastrear envío', error);
    }
  }
  
  // ==================== RECOLECCIONES ====================
  
  /**
   * Obtiene cobertura de fechas para recolección
   * @param {string} shipmentId - ID del envío
   * @returns {Promise<Object>} Fechas disponibles
   */
  async getPickupCoverage(shipmentId) {
    try {
      const response = await this.httpClient.get('/api/v1/pickups/coverage', {
        params: { shipment_id: shipmentId }
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener cobertura de recolección', error);
    }
  }
  
  /**
   * Crea una recolección
   * @param {Object} pickupData - Datos de la recolección
   * @returns {Promise<Object>} Recolección creada
   * 
   * @example
   * const pickup = await client.createPickup({
   *   reference_shipment_id: 'shipment_id',
   *   packages: 1,
   *   total_weight: 2.5,
   *   scheduled_from: '2024-01-15T09:00:00.000-06:00',
   *   scheduled_to: '2024-01-15T14:00:00.000-06:00'
   * });
   */
  async createPickup(pickupData) {
    try {
      const response = await this.httpClient.post('/api/v1/pickups', {
        pickup: pickupData
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al crear recolección', error);
    }
  }
  
  /**
   * Obtiene todas las recolecciones (paginadas)
   * @param {number} [page=1] - Número de página
   * @returns {Promise<Object>} Lista de recolecciones
   */
  async getPickups(page = 1) {
    try {
      const response = await this.httpClient.get('/api/v1/pickups', {
        params: { page }
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener recolecciones', error);
    }
  }
  
  /**
   * Obtiene una recolección por ID
   * @param {string} pickupId - ID de la recolección
   * @returns {Promise<Object>} Detalles de la recolección
   */
  async getPickup(pickupId) {
    try {
      const response = await this.httpClient.get(`/api/v1/pickups/${pickupId}`);
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener recolección', error);
    }
  }
  
  /**
   * Reprograma una recolección
   * @param {Object} pickupData - Datos de reprogramación
   * @returns {Promise<Object>} Recolección reprogramada
   */
  async reschedulePickup(pickupData) {
    try {
      const response = await this.httpClient.post('/api/v1/pickups/reschedule', {
        pickup: pickupData
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al reprogramar recolección', error);
    }
  }
  
  // ==================== ÓRDENES ====================
  
  /**
   * Obtiene todas las órdenes
   * @returns {Promise<Object>} Lista de órdenes
   */
  async getOrders() {
    try {
      const response = await this.httpClient.get('/api/v1/orders');
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener órdenes', error);
    }
  }
  
  /**
   * Obtiene una orden por ID
   * @param {string} orderId - ID de la orden
   * @returns {Promise<Object>} Detalles de la orden
   */
  async getOrder(orderId) {
    try {
      const response = await this.httpClient.get(`/api/v1/orders/${orderId}`);
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener orden', error);
    }
  }
  
  /**
   * Crea una orden
   * @param {Object} orderData - Datos de la orden
   * @returns {Promise<Object>} Orden creada
   */
  async createOrder(orderData) {
    try {
      const response = await this.httpClient.post('/api/v1/orders', {
        order: orderData
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al crear orden', error);
    }
  }
  
  /**
   * Actualiza una orden
   * @param {string} orderId - ID de la orden
   * @param {Object} orderData - Datos actualizados
   * @returns {Promise<Object>} Orden actualizada
   */
  async updateOrder(orderId, orderData) {
    try {
      const response = await this.httpClient.patch(`/api/v1/orders/${orderId}`, {
        order: orderData
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al actualizar orden', error);
    }
  }
  
  // ==================== PRODUCTOS ====================
  
  /**
   * Obtiene lista de productos (paginados)
   * @param {number} [page=1] - Número de página
   * @param {Object} [filters={}] - Filtros opcionales
   * @returns {Promise<Object>} Lista de productos
   */
  async getProducts(page = 1, filters = {}) {
    try {
      const response = await this.httpClient.get('/api/v1/products', {
        params: { page, ...filters }
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError('Error al obtener productos', error);
    }
  }
  
  // ==================== CONFIGURACIÓN ====================
  
  /**
   * Actualiza el formato de impresión
   * @param {string} format - Formato: 'standard' o 'thermal'
   * @returns {Promise<void>}
   */
  async updatePrintingFormat(format) {
    try {
      await this.httpClient.patch('/api/v1/settings/printing_formats', {
        printing_format: format
      });
    } catch (error) {
      throw this.handleError('Error al actualizar formato de impresión', error);
    }
  }
  
  // ==================== UTILIDADES ====================
  
  /**
   * Maneja errores de la API
   * @private
   */
  handleError(message, error) {
    const errorInfo = {
      message,
      status: error.response?.status,
      statusText: error.response?.statusText,
      data: error.response?.data,
      originalError: error.message
    };
    
    const errorMessage = `${message}: ${error.response?.data?.message || error.message}`;
    const apiError = new Error(errorMessage);
    apiError.details = errorInfo;
    
    return apiError;
  }
  
  /**
   * Obtiene información del cliente
   * @returns {Object} Información de configuración
   */
  getClientInfo() {
    return {
      environment: this.environment,
      baseURL: this.baseURL,
      hasValidToken: !!this.accessToken && Date.now() < this.tokenExpiresAt,
      tokenExpiresAt: this.tokenExpiresAt ? new Date(this.tokenExpiresAt) : null
    };
  }
}

module.exports = SkydropxClient;

File: src/clients/python/skydropx_client.py
"""
Skydropx API Client for Python

Este cliente permite interactuar con la API de Skydropx de manera sencilla.
Incluye autenticación automática, renovación de tokens y manejo de errores.

Uso básico:
    from skydropx_client import SkydropxClient
    
    client = SkydropxClient(
        client_id='your_client_id',
        client_secret='your_client_secret',
        environment='sandbox'
    )
    
    # Autenticar
    client.authenticate()
    
    # Crear cotización
    quotation = client.create_quotation({
        'address_from': {...},
        'address_to': {...},
        'packages': [...]
    })
"""

import requests
import time
import json
from typing import Dict, List, Optional, Any
from datetime import datetime, timedelta


class SkydropxError(Exception):
    """Excepción base para errores de Skydropx"""
    
    def __init__(self, message: str, status_code: Optional[int] = None, response_data: Optional[Dict] = None):
        self.message = message
        self.status_code = status_code
        self.response_data = response_data
        super().__init__(self.message)


class SkydropxClient:
    """
    Cliente para la API de Skydropx
    
    Args:
        client_id: Client ID de OAuth
        client_secret: Client Secret de OAuth
        environment: 'sandbox' o 'production'
        auto_renew_token: Si debe renovar automáticamente el token (default: True)
    """
    
    BASE_URLS = {
        'sandbox': 'https://app.skydropx.com',
        'production': 'https://app.skydropx.com'
    }
    
    def __init__(
        self,
        client_id: str,
        client_secret: str,
        environment: str = 'sandbox',
        auto_renew_token: bool = True
    ):
        self.client_id = client_id
        self.client_secret = client_secret
        self.environment = environment
        self.auto_renew_token = auto_renew_token
        
        self.base_url = self.BASE_URLS.get(environment, self.BASE_URLS['sandbox'])
        self.access_token: Optional[str] = None
        self.token_expires_at: Optional[datetime] = None
        
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'Skydropx-Python-SDK/1.0.0'
        })
    
    def _should_renew_token(self) -> bool:
        """Verifica si el token debe renovarse"""
        if not self.access_token or not self.token_expires_at:
            return True
        
        # Renovar 5 minutos antes de expirar
        return datetime.now() >= (self.token_expires_at - timedelta(minutes=5))
    
    def _handle_error(self, response: requests.Response) -> None:
        """Maneja errores de la API"""
        try:
            error_data = response.json()
        except:
            error_data = {'error': response.text}
        
        error_messages = {
            400: 'Solicitud inválida',
            401: 'No autorizado - Verifica tus credenciales',
            403: 'Acceso prohibido',
            404: 'Recurso no encontrado',
            422: 'Error de validación',
            429: 'Límite de tasa excedido - Intenta más tarde',
            500: 'Error interno del servidor',
            503: 'Servicio no disponible'
        }
        
        message = error_messages.get(response.status_code, f'Error {response.status_code}')
        
        raise SkydropxError(
            message=message,
            status_code=response.status_code,
            response_data=error_data
        )
    
    def _request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict] = None,
        params: Optional[Dict] = None,
        requires_auth: bool = True
    ) -> Dict:
        """Realiza una petición a la API"""
        
        # Renovar token si es necesario
        if requires_auth and self.auto_renew_token and self._should_renew_token():
            self.authenticate()
        
        # Configurar headers
        headers = {}
        if requires_auth and self.access_token:
            headers['Authorization'] = f'Bearer {self.access_token}'
        
        # Realizar petición
        url = f"{self.base_url}{endpoint}"
        
        try:
            response = self.session.request(
                method=method,
                url=url,
                json=data,
                params=params,
                headers=headers,
                timeout=30
            )
            
            if not response.ok:
                self._handle_error(response)
            
            return response.json() if response.content else {}
            
        except requests.exceptions.Timeout:
            raise SkydropxError('Timeout - La solicitud tardó demasiado')
        except requests.exceptions.ConnectionError:
            raise SkydropxError('Error de conexión - Verifica tu internet')
        except SkydropxError:
            raise
        except Exception as e:
            raise SkydropxError(f'Error inesperado: {str(e)}')
    
    # ============= AUTENTICACIÓN =============
    
    def authenticate(self) -> Dict:
        """
        Obtiene un access token de OAuth
        
        Returns:
            Dict con información del token
        """
        data = {
            'client_id': self.client_id,
            'client_secret': self.client_secret,
            'grant_type': 'client_credentials'
        }
        
        response = self._request(
            'POST',
            '/api/v1/oauth/token',
            data=data,
            requires_auth=False
        )
        
        self.access_token = response['access_token']
        expires_in = response.get('expires_in', 7200)
        self.token_expires_at = datetime.now() + timedelta(seconds=expires_in)
        
        return response
    
    def revoke_token(self) -> Dict:
        """Revoca el token actual"""
        data = {
            'client_id': self.client_id,
            'client_secret': self.client_secret,
            'token': self.access_token,
            'token_type_hint': 'access_token'
        }
        
        response = self._request(
            'POST',
            '/api/v1/oauth/revoke',
            data=data,
            requires_auth=False
        )
        
        self.access_token = None
        self.token_expires_at = None
        
        return response
    
    def introspect_token(self) -> Dict:
        """Obtiene información del token actual"""
        data = {
            'client_id': self.client_id,
            'client_secret': self.client_secret,
            'token': self.access_token,
            'token_type_hint': 'access_token'
        }
        
        return self._request(
            'POST',
            '/api/v1/oauth/introspect',
            data=data,
            requires_auth=False
        )
    
    # ============= COTIZACIONES =============
    
    def create_quotation(self, quotation_data: Dict) -> Dict:
        """
        Crea una cotización
        
        Args:
            quotation_data: Datos de la cotización
            
        Returns:
            Dict con información de la cotización
        """
        return self._request(
            'POST',
            '/api/v1/quotations',
            data={'quotation': quotation_data}
        )
    
    def get_quotation(self, quotation_id: str) -> Dict:
        """
        Obtiene los resultados de una cotización
        
        Args:
            quotation_id: ID de la cotización
            
        Returns:
            Dict con tarifas disponibles
        """
        return self._request(
            'GET',
            f'/api/v1/quotations/{quotation_id}'
        )
    
    def wait_for_quotation(self, quotation_id: str, max_attempts: int = 15, sleep_seconds: int = 2) -> Dict:
        """
        Espera a que una cotización se complete (polling)
        
        Args:
            quotation_id: ID de la cotización
            max_attempts: Número máximo de intentos
            sleep_seconds: Segundos entre intentos
            
        Returns:
            Dict con cotización completada
        """
        attempts = 0
        
        while attempts < max_attempts:
            quotation = self.get_quotation(quotation_id)
            
            if quotation.get('is_completed'):
                return quotation
            
            time.sleep(sleep_seconds)
            attempts += 1
        
        raise SkydropxError('Timeout esperando cotización - Intenta más tarde')
    
    # ============= ENVÍOS =============
    
    def create_shipment(self, shipment_data: Dict) -> Dict:
        """
        Crea un envío
        
        Args:
            shipment_data: Datos del envío
            
        Returns:
            Dict con información del envío
        """
        return self._request(
            'POST',
            '/api/v1/shipments',
            data={'shipment': shipment_data}
        )
    
    def get_shipments(self, params: Optional[Dict] = None) -> Dict:
        """
        Lista envíos con filtros opcionales
        
        Args:
            params: Filtros (page, per_page, status, etc)
            
        Returns:
            Dict con lista de envíos
        """
        return self._request(
            'GET',
            '/api/v1/shipments',
            params=params
        )
    
    def get_shipment(self, shipment_id: str) -> Dict:
        """
        Obtiene un envío por ID
        
        Args:
            shipment_id: ID del envío
            
        Returns:
            Dict con información del envío
        """
        return self._request(
            'GET',
            f'/api/v1/shipments/{shipment_id}'
        )
    
    def cancel_shipment(self, shipment_id: str, reason: str = '') -> Dict:
        """
        Cancela un envío
        
        Args:
            shipment_id: ID del envío
            reason: Razón de cancelación
            
        Returns:
            Dict con envío cancelado
        """
        return self._request(
            'POST',
            f'/api/v1/shipments/{shipment_id}/cancel',
            data={'cancellation_reason': reason}
        )
    
    def protect_shipment(self, shipment_id: str, declared_value: float) -> Dict:
        """
        Agrega seguro a un envío
        
        Args:
            shipment_id: ID del envío
            declared_value: Valor declarado
            
        Returns:
            Dict con envío asegurado
        """
        return self._request(
            'POST',
            f'/api/v1/shipments/{shipment_id}/protect',
            data={'declared_value': declared_value}
        )
    
    # ============= RASTREO =============
    
    def track_shipment(self, tracking_number: str, carrier_code: str) -> Dict:
        """
        Rastrea un envío
        
        Args:
            tracking_number: Número de guía
            carrier_code: Código de paquetería (fedex, dhl, etc)
            
        Returns:
            Dict con información de rastreo
        """
        return self._request(
            'GET',
            '/api/v1/tracking',
            params={
                'tracking_number': tracking_number,
                'carrier_code': carrier_code
            }
        )
    
    def track_multiple_shipments(self, trackings: List[Dict]) -> Dict:
        """
        Rastrea múltiples envíos
        
        Args:
            trackings: Lista de dicts con tracking_number y carrier_code
            
        Returns:
            Dict con información de múltiples rastreos
        """
        return self._request(
            'POST',
            '/api/v1/tracking/bulk',
            data={'trackings': trackings}
        )
    
    # ============= RECOLECCIONES =============
    
    def get_pickup_coverage(self, postal_code: str, country_code: str = 'MX') -> Dict:
        """
        Verifica cobertura de recolección
        
        Args:
            postal_code: Código postal
            country_code: Código de país
            
        Returns:
            Dict con información de cobertura
        """
        return self._request(
            'POST',
            '/api/v1/pickup_coverage',
            data={
                'zip': postal_code,
                'country_code': country_code
            }
        )
    
    def create_pickup(self, pickup_data: Dict) -> Dict:
        """
        Programa una recolección
        
        Args:
            pickup_data: Datos de la recolección
            
        Returns:
            Dict con información de la recolección
        """
        return self._request(
            'POST',
            '/api/v1/pickups',
            data={'pickup': pickup_data}
        )
    
    def get_pickups(self, params: Optional[Dict] = None) -> Dict:
        """
        Lista recolecciones
        
        Args:
            params: Filtros opcionales
            
        Returns:
            Dict con lista de recolecciones
        """
        return self._request(
            'GET',
            '/api/v1/pickups',
            params=params
        )
    
    def reschedule_pickup(self, pickup_id: str, pickup_data: Dict) -> Dict:
        """
        Reprograma una recolección
        
        Args:
            pickup_id: ID de la recolección
            pickup_data: Nuevos datos
            
        Returns:
            Dict con recolección actualizada
        """
        return self._request(
            'PUT',
            f'/api/v1/pickups/{pickup_id}/reschedule',
            data=pickup_data
        )
    
    # ============= WEBHOOKS =============
    
    def create_webhook(self, webhook_data: Dict) -> Dict:
        """
        Registra un webhook
        
        Args:
            webhook_data: Datos del webhook
            
        Returns:
            Dict con información del webhook
        """
        return self._request(
            'POST',
            '/api/v1/webhooks',
            data={'webhook': webhook_data}
        )
    
    def get_webhooks(self) -> Dict:
        """
        Lista webhooks registrados
        
        Returns:
            Dict con lista de webhooks
        """
        return self._request(
            'GET',
            '/api/v1/webhooks'
        )
    
    def update_webhook(self, webhook_id: str, webhook_data: Dict) -> Dict:
        """
        Actualiza un webhook
        
        Args:
            webhook_id: ID del webhook
            webhook_data: Datos actualizados
            
        Returns:
            Dict con webhook actualizado
        """
        return self._request(
            'PUT',
            f'/api/v1/webhooks/{webhook_id}',
            data={'webhook': webhook_data}
        )
    
    def delete_webhook(self, webhook_id: str) -> Dict:
        """
        Elimina un webhook
        
        Args:
            webhook_id: ID del webhook
            
        Returns:
            Dict vacío en caso de éxito
        """
        return self._request(
            'DELETE',
            f'/api/v1/webhooks/{webhook_id}'
        )
    
    # ============= UTILIDADES =============
    
    def get_client_info(self) -> Dict:
        """
        Obtiene información del cliente
        
        Returns:
            Dict con información de configuración
        """
        return {
            'environment': self.environment,
            'base_url': self.base_url,
            'has_valid_token': bool(self.access_token and not self._should_renew_token()),
            'token_expires_at': self.token_expires_at.isoformat() if self.token_expires_at else None
        }


def verify_webhook_signature(signature: str, timestamp: str, payload: str, secret: str) -> bool:
    """
    Verifica la firma HMAC de un webhook
    
    Args:
        signature: Firma del header X-Skydropx-Signature
        timestamp: Timestamp del header X-Skydropx-Timestamp
        payload: Body del request como string
        secret: Secret del webhook
        
    Returns:
        True si la firma es válida
    """
    import hmac
    import hashlib
    
    # Crear payload firmado
    signed_payload = f"{timestamp}.{payload}"
    
    # Calcular HMAC-SHA512
    expected_signature = hmac.new(
        secret.encode(),
        signed_payload.encode(),
        hashlib.sha512
    ).hexdigest()
    
    # Comparar (time-safe)
    signature_without_prefix = signature.replace('sha512=', '')
    
    return hmac.compare_digest(signature_without_prefix, expected_signature)
